### PR TITLE
build/ops: conditionalize rgw Beast frontend so it isn't built on s390x architecture

### PR DESCRIPTION
**Pull request by smithfarm**
Fix for http://tracker.ceph.com/issues/20048

Problem solved by this PR: boost::component is not supported on s390x architecture, breaking the s390x build.

Also, it's a Good Idea in general to tie dependencies to the code that actually uses them.

(N.B.: Ceph is being built on s390x only so Linux VMs on the mainframe can function as Ceph clients.)

### CODE REVIEW

**Comment by smithfarm** [comment]
https://shaman.ceph.com/builds/ceph/wip-s390-rgw/

**Comment by smithfarm** [comment]
Confirmed that this doesn't break any of the upstream builds. Now trying to get it to build successfully on s390x.

**Comment by cbodley** [comment]
can you paste your `build/CMakeFiles/CMakeError.log` when it fails to find boost context?

**Comment by smithfarm** [comment]
@cbodley I can try, but I currently have no way to access the remote build environment where these s390x builds are taking place.

File: CMakeLists.txt [hunk 1]
```diff
@@ -512,13 +512,17 @@ endif()
...
 set(BOOST_COMPONENTS
-	thread system regex random program_options date_time iostreams coroutine context)
+	thread system regex random program_options date_time iostreams coroutine)
 set(BOOST_HEADER_COMPONENTS container)
 
 if(WITH_MGR)
 	list(APPEND BOOST_COMPONENTS python)
 endif()
 
+if(WITH_RADOSGW)
```
Comment (cbodley) [inline CMakeLists.txt]: `if (WITH_RADOSGW_BEAST_FRONTEND)` should be sufficient here. then you don't need to change the packaging for radosgw

**Comment by mdw-at-linuxbox** [comment]
Not being able to see error logs from s390x seems to me to be a problem in itself.

Also it would be nice to know more about why this was a problem on s390x - does this mean we have to avoid using boost::context everywhere else in ceph too?

**Comment by tserong** [comment]
> Not being able to see error logs from s390x seems to me to be a problem in itself.

That's because it's built in a build farm, where users don't have access to the insides of the VMs where the builds are done.

> Also it would be nice to know more about why this was a problem on s390x - does this mean we have to avoid using boost::context everywhere else in ceph too?

Apparently boost::context isn't supported on s390x.  From https://github.com/SUSE/ceph/pull/114#issuecomment-302806072:

````
[  121s] libs/context/src/unsupported.cpp:7:2: error: #error "platform not supported"
[  121s]  #error "platform not supported"
[  121s]   ^
[  164s] ...skipped <p/home/abuild/rpmbuild/BUILD/ceph-12.0.3+git.1495217527.9b626e3f69/build/boost-build/boost/bin.v2/libs/context/build/gcc-4.8/release/link-static/threading-multi>libboost_context.a(clean) for lack of <p/home/abuild/rpmbuild/BUILD/ceph-12.0.3+git.1495217527.9b626e3f69/build/boost-build/boost/bin.v2/libs/context/build/gcc-4.8/release/link-static/threading-multi>unsupported.o...
[  164s] ...skipped <p/home/abuild/rpmbuild/BUILD/ceph-12.0.3+git.1495217527.9b626e3f69/build/boost-build/boost/bin.v2/libs/context/build/gcc-4.8/release/link-static/threading-multi>libboost_context.a for lack of <p/home/abuild/rpmbuild/BUILD/ceph-12.0.3+git.1495217527.9b626e3f69/build/boost-build/boost/bin.v2/libs/context/build/gcc-4.8/release/link-static/threading-multi>unsupported.o...
[  164s] ...skipped <p/home/abuild/rpmbuild/BUILD/ceph-12.0.3+git.1495217527.9b626e3f69/build/boost/lib>libboost_context.a for lack of <p/home/abuild/rpmbuild/BUILD/ceph-12.0.3+git.1495217527.9b626e3f69/build/boost-build/boost/bin.v2/libs/context/build/gcc-4.8/release/link-static/threading-multi>libboost_context.a...
[  196s] ...failed updating 1 target...
[  196s] CMake Warning at /usr/share/cmake/Modules/FindBoost.cmake:725 (message):
[  196s]   Imported targets not available for Boost version 106300
[  196s] Call Stack (most recent call first):
[  196s]   /usr/share/cmake/Modules/FindBoost.cmake:763 (_Boost_COMPONENT_DEPENDENCIES)
[  196s]   /usr/share/cmake/Modules/FindBoost.cmake:1315 (_Boost_MISSING_DEPENDENCIES)
[  196s]   CMakeLists.txt:594 (find_package)
````

**Comment by smithfarm** [comment]
Please see the description of http://tracker.ceph.com/issues/20048 for my analysis of the build failure.

My guess is that context is indeed not supported on s390x and, yes, this would mean that it should either be avoided or used with care so as not to break the s390x build.

**Comment by cbodley** [comment]
@tserong @smithfarm thanks, that's good to know

i do have plans to expand the use of `boost::asio` in radosgw outside of the beast frontend (and potentially into the `Objecter`), but i'll be cautious about using the stackful coroutines that depend on `boost::context`

File: CMakeLists.txt [hunk 2]
```diff
@@ -512,13 +512,17 @@ endif()
 option(WITH_SYSTEM_BOOST "require and build with system Boost" OFF)
 
 set(BOOST_COMPONENTS
-	thread system regex random program_options date_time iostreams coroutine context)
+	thread system regex random program_options date_time iostreams coroutine)
```
Comment (cbodley) [inline CMakeLists.txt]: the `coroutine` component is also specific to the beast frontend

**Comment by smithfarm** [comment]
@cbodley Thanks for understanding. Coroutine moved; rebased.

**Comment by smithfarm** [comment]
@cbodley I added the debian bit. We are using this PR now downstream. Would be great to get it merged, if you agree.

@tchaikov Can you review?

File: CMakeLists.txt [hunk 3]
```diff
@@ -512,13 +512,17 @@ endif()
...
 set(BOOST_COMPONENTS
-	thread system regex random program_options date_time iostreams coroutine context)
+	thread system regex random program_options date_time iostreams)
 set(BOOST_HEADER_COMPONENTS container)
 
 if(WITH_MGR)
 	list(APPEND BOOST_COMPONENTS python)
 endif()
 
+if(WITH_RADOSGW_BEAST_FRONTEND)
```
Comment (tchaikov) [inline CMakeLists.txt]: we should disable `WITH_RADOSGW_BEAST_FRONTEND` archs not listed below :
- arm32
- aarch64
- i386
- amd64
- powerpc32
- powerpc64
- mips32

i am consulting `asm_context_sources` in `boost/libs/context/build/Jamfile.v2` for the above list.

in other words, following archs are not supported:

- alpha
- hppa
- ia64
- m68k
- mips64
- mips64el
- s390
- s390x
- sh4
- sparc
- sparc64
- x32

i am consulting `boost1.63-1.63.0+dfsg/debian/rules` for the list above.

considering the length of these lists and their authorities, probably we should use the former.

File: debian/rules [hunk 4]
```diff
@@ -22,6 +22,10 @@ ifeq ($(DEB_HOST_ARCH), armel)
   extraopts += -DWITH_ATOMIC_OPS=OFF
 endif
 
+ifeq (,$(filter $(DEB_HOST_ARCH), s390 s390x))
```
Comment (tchaikov) [inline debian/rules]: should be `ifneq`

Comment (smithfarm) [inline debian/rules]: the conditional is supposed to be: "if DEB_HOST_ARCH equals "s390" or "s390x", then build with "-DWITH_RADOSGW_BEAST_FRONTEND=OFF"

I've never coded in this language, so I take your word for it that `ifneq` is correct in this case.

File: CMakeLists.txt (hunk 3 above)
Comment (smithfarm) [inline CMakeLists.txt]: Thanks for looking that up - I can change the conditionals if you wish, though I doubt anyone will be building RPM and deb packages for mips32, for example.

File: debian/rules (hunk 4 above)
Comment (tchaikov) [inline debian/rules]: @smithfarm see https://www.gnu.org/software/make/manual/html_node/Text-Functions.html. if it matches, a `s390` or `s390x` is returned, which is non-empty. in that case we should pass `-DWITH_RADOSGW_BEAST_FRONTEND=OFF` to cmake.

File: CMakeLists.txt (hunk 3 above)
Comment (tchaikov) [inline CMakeLists.txt]: @smithfarm debian supports mips and mipsel, so for the sake of completeness, we can list it in `debian/rules` like
```makefile
ifeq (,$(filter $(DEB_HOST_ARCH), arm armel armhf arm64 i386 amd64 mips mipsel powerpc ppc64))
  extraopts += -DWITH_RADOSGW_BEAST_FRONTEND=ON
else
  extraopts += -DWITH_RADOSGW_BEAST_FRONTEND=OFF
endif
```

i understand if you intent to keep clean and minimal. but i just wanted to the packaging to be a better reference for our downstreams or potential users.

**Comment by tchaikov** [comment]
see also #15346, @smithfarm i can help with the `debian/rules`, if you'd like.

**Comment by tchaikov** [comment]
see also #15347

**Comment by tchaikov** [comment]
@smithfarm could you squash related commits into a single one? and remove the merge commit?

**Comment by smithfarm** [comment]
@tchaikov Yes, I'm working on that. What about "cmake: use boost target instead libary path while linking against them" and "cmake: delete excidentail reintroduced line." ? I guess we need those as well?

**Comment by smithfarm** [comment]
https://github.com/ceph/ceph/pull/15346/commits/8eb58215fc240dbff392bd65f4a482f2833021d0 has been deleted in tchaikov/ceph.git ?

**Comment by smithfarm** [comment]
@tchaikov Please have another look. I guess this now obsoletes #15347, too.

**Comment by tchaikov** [comment]
@smithfarm 974edd9 is more of a cosmetic change, and it fails the build in my last PR. FindBoost sometimes fails to find the built boost components. i am not sure why, though.

**Comment by smithfarm** [comment]
@tchaikov OK, dropped that commit and adapted the first commit to match #15347

**Comment by tchaikov** [comment]
@smithfarm i thought you included #15347 in this PR. but seems that's not the case. so i reopened it.

**Comment by smithfarm** [comment]
@tchaikov I did, but I can't get it to build without the wholesale Boost linking.

**Comment by smithfarm** [comment]
sorry, I meant "I can't get it to build _with_ the wholesale Boost linking"

**Comment by smithfarm** [comment]
@cbodley @ktdreyer This is close to being merged; wanna take a look?

**Comment by smithfarm** [comment]
Status update: This passes Shaman, but still fails in OBS (apparently because radosgw is not explicitly linked with Boost). I'm testing a fix for that now.

**Comment by smithfarm** [comment]
https://shaman.ceph.com/builds/ceph/wip-s390-rgw/3dfffa59961856d2b220601964e2816c7cd52678/

**Comment by smithfarm** [comment]
OBS build is also green: https://build.opensuse.org/package/show/filesystems:ceph:luminous/ceph

**Comment by smithfarm** [comment]
s390x build is green. AFAICT this is ready to merge.

**Pull request merged**